src/Storage/Notmuch.hs

```diff
@@ -49,22 +51,22 @@ setNotmuchMailTags dbpath m =
   case mailTagsToNotmuchTags m of
     Nothing -> pure $ Left "Tags are corrupt"
     Just nmtags ->
-      bracket (runExceptT (databaseOpen dbpath) >>= either (error . show) pure)
-              (void . runExceptT . databaseDestroy)
-              (\db -> tagsToMessage nmtags m db)
+      bracket (runExceptT (databaseOpen dbpath))
+              (runExceptT . traverse_ databaseDestroy)
+              (runExceptT . (either (throwError . show) pure >=> tagsToMessage nmtags m))
```

Comment:
Inside the monad transformer you propagate the error by "rethrowing" it?

Reply:
Essentially yes, because we want to work in MonadError, but `bracket` forces us to `IO`.  So for each "stage" of `bracket` we `runExceptT`, and then in the "main computation" stage we
take the `Either e a` from the resource initialisation stage and "lift" it back into a `MonadError e IO a` to continue the computation.

After I tackle the "new error sum type" refactor discussed in the comment later in commit,
I'll factor out a `MonadError`-friendly variant of `bracket` that lets us avoid doing all of this juggling between `IO (Either e a)` and `MonadError e IO a`.

Reply:
I understand. So merge this and the sum type refactor comes with different patches and separate commits in hs-notmuch, yes?

Reply:
Yep.  I'll take that as your ACK :)